examples/performance-debugging/performance-debugging.cpp

```diff
@@ -0,0 +1,500 @@
+                               const gko::LinOp *residual_norm) const override
+    {
+        if (residual_norm) {
+            rec_res_norms.push_back(
+                utils::get_norm(gko::as<vec<ValueType>>(residual_norm)));
+        } else {
+            rec_res_norms.push_back(
+                utils::compute_norm(gko::as<vec<ValueType>>(residual)));
+        }
+        if (solution) {
+            true_res_norms.push_back(utils::compute_residual_norm(
+                matrix, b, gko::as<vec<ValueType>>(solution)));
+        } else {
+            true_res_norms.push_back(-1.0);
+        }
+    }
+
+    ResidualLogger(std::shared_ptr<const gko::Executor> exec,
+                   const gko::LinOp *matrix, const vec<ValueType> *b)
+        : gko::log::Logger(exec, gko::log::Logger::iteration_complete_mask),
+          matrix{matrix},
+          b{b}
+    {}
+
+    void write_data(std::ostream &ostream)
+    {
+        ostream << "Recurrent Residual Norms: " << std::endl;
+        ostream << "[" << std::endl;
+        for (const auto &entry : rec_res_norms) {
+            ostream << "\t" << entry << std::endl;
+        }
+        ostream << "];" << std::endl;
+
+        ostream << "True Residual Norms: " << std::endl;
+        ostream << "[" << std::endl;
+        for (const auto &entry : true_res_norms) {
+            ostream << "\t" << entry << std::endl;
+        }
+        ostream << "];" << std::endl;
+    }
+
+private:
+    const gko::LinOp *matrix;
+    const vec<ValueType> *b;
+    mutable std::vector<ValueType> rec_res_norms;
+    mutable std::vector<ValueType> true_res_norms;
+};
+
+
+}  // namespace loggers
+
+
+namespace {
+
+
+// Print usage help
+void print_usage(const char *filename)
+{
+    std::cerr << "Usage: " << filename << " [executor] [matrix file]"
+              << std::endl;
+    std::cerr << "matrix file should be a file in matrix market format. "
+                 "The file data/A.mtx is provided as an example."
+              << std::endl;
+    std::exit(-1);
+}
+
+
+void print_vector(const gko::matrix::Dense<> *vec)
+{
+    auto elements_to_print = std::min(gko::size_type(10), vec->get_size()[0]);
+    std::cout << "[" << std::endl;
+    for (int i = 0; i < elements_to_print; ++i) {
+        std::cout << "\t" << vec->at(i) << std::endl;
+    }
+    std::cout << "];" << std::endl;
+}
+
+
+}  // namespace
+
+
+int main(int argc, char *argv[])
+{
+    // Parametrize the benchmark here
+    // Pick a value type
+    using vtype = double;
+    // Pick a matrix format
+    using mtx = gko::matrix::Csr<vtype>;
+    // Pick a solver
+    using solver = gko::solver::Cg<vtype>;
+    // Pick a preconditioner type
+    using preconditioner = gko::matrix::IdentityFactory<vtype>;
+    // Pick a residual norm reduction value
+    auto reduction_factor = 1e-8;
+    // Pick a maximum iteration count
+    auto max_iters = 2000u;
+    // Pick an output file name
+    auto of_name = "data.txt";
```

Comment:
I prefer to print information to stdout and use redirection to a file.

Reply:
The problem with this is that you have the ginkgo version related lines and extra unrelated information which clutters the results. In addition, if you have a big residual etc, printing to screen will give you no information due to the amout of lines you would need to scroll through in the console. Should I proceed with this nonetheless? What do you think @hartwiganzt?

Reply:
I see Mike's idea: by default print everything on screen, if it is too much, redirect the output to a file. I don't have a strong opinion here, is there a disadvantage of creating a "data.txt" file for selected output?

Reply:
I would actually prefer the (current) file version because:
1. Since it is a logger you would actually expect a log file of some sort (maybe `log` should be in the default name)
2. There are other outputs (Ginkgo version, result output), so redirecting to a file will contain data unrelated to the logging (very minor argument)
3. The logging can get quite large, especially the "Recurrent" and "True Residual Norms", which are harder to see if it is just stdout

I also have no problem if we put it to stdout, but I think as an example and for debugging or logging, it will always end up in a file, which is why I would prefer doing it directly.

Reply:
I see. Putting the logging in the file is okay for me